<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make sashimi-esque plot with ggplot2 — make_differential_splicing_plot • leafcutter</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">leafcutter</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/Installation.html">Installation</a>
</li>
<li>
  <a href="../articles/Usage.html">Differential splicing</a>
</li>
<li>
  <a href="../articles/sQTL.html">SplicingQTL</a>
</li>
<li>
  <a href="../reference/index.html">R functions</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/davidaknowles/leafcutter">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make sashimi-esque plot with ggplot2</h1>
    </div>

    
    <p>Shows only junction reads and can optionally show splicing variation across groups.</p>
    

    <pre class="usage"><span class='fu'>make_differential_splicing_plot</span>(<span class='no'>y</span>, <span class='kw'>x</span> <span class='kw'>=</span> <span class='fu'>numeric</span>(<span class='fu'>nrow</span>(<span class='no'>y</span>)) + <span class='fl'>1</span>,
  <span class='kw'>exons_table</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>len</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>length_transform</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>g</span>) <span class='fu'>log</span>(<span class='no'>g</span> +
  <span class='fl'>1</span>), <span class='kw'>main_title</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>snp_pos</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>summary_func</span> <span class='kw'>=</span> <span class='no'>colMeans</span>,
  <span class='kw'>legend_title</span> <span class='kw'>=</span> <span class='st'>"Mean count"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p>[samples] x [introns] matrix of counts</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>[samples] vector of group membership. Can be numeric, factor or character.</p></td>
    </tr>
    <tr>
      <th>exons_table</th>
      <td><p>An optional data frame specifying exons, with columns: chr start end strand gene_name. For hg19 see data/gencode19_exons.txt.gz</p></td>
    </tr>
    <tr>
      <th>len</th>
      <td><p>Number of segments each curve is constructed from, controls smoothness.</p></td>
    </tr>
    <tr>
      <th>length_transform</th>
      <td><p>Function controlling how true genomic space is mapped to the plot for improved visability. Use the identity function (i.e. function(g) g</p></td>
    </tr>
    <tr>
      <th>main_title</th>
      <td><p>Title</p></td>
    </tr>
    <tr>
      <th>snp_pos</th>
      <td><p>An optional list of SNP positions to mark</p></td>
    </tr>
    <tr>
      <th>summary_func</th>
      <td><p>Function to summarize counts within groups: usually colMeans or colSums</p></td>
    </tr>
    <tr>
      <th>legend_title</th>
      <td><p>Match this to summary_func</p></td>
    </tr>
    </table>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
                </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Knowles, Yang Li, Jonathan Pritchard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
